lib/elixir/lib/kernel.ex

```diff
@@ -4295,6 +4295,65 @@ defmodule Kernel do
   end
 
   @doc """
+  Generates a macro suitable for use in guard expressions.
+
+  It raises at compile time if the definition uses expressions that aren't
+  allowed in guards, and otherwise creates a macro that can be used both inside
+  or outside guards.
+
+  ## Example
+
+      defmodule Integer.Guards do
+        defguard is_even(value) when is_integer(value) and rem(value, 2) == 0
+      end
+
+      defmodule Collatz do
+        @moduledoc "Tools for working with the Collatz Conjecture."
+        import Integer.Guards
+
+        @doc "Determines the number of steps `n` takes to reach `1`."
+        def steps(n) when n > 0, do: steps(n, 0)
+
+        defp step(1, step_count), do: step_count
+        defp step(n, step_count) when is_even(n), do: step(div(n, 2), step_count + 1)
+        defp step(n, step_count), do: step(3*n + 1, step_count + 1)
+      end
+
+  """
+  @spec defguard(Macro.t) :: Macro.t | no_return
+  defmacro defguard(guard) do
+    {call, impl} = :elixir_utils.extract_guards(guard)
+    case Macro.decompose_call(call) do
+      {name, args} ->
+        do_defguard(call, name, args, impl, __CALLER__)
+      _invalid_definition ->
+        raise ArgumentError, message: "invalid syntax in defguard #{Macro.to_string call}"
+    end
+  end
+
+  defp do_defguard(_call, name, args, _impl = [], env) do
+    message = :elixir_def.format_error({:bodyless_clause, :defguard, {name, length(args)}})
```

Comment:
I would say we should just expand this case to:

```elixir
defmacro unquote(call)
```

And Elixir will take care of emitting the warning if no implementation is defined later.